### ISSUE TITLE
Add reset to macro context

### ISSUE THREAD

**Issue by disnet** [sweet-js/sweet.js#537]
As discussed in #529 we need a way to reset the macro context. A `.reset` is the most obvious:

```js
syntax m = ctx => {
  ctx.next('expr'); // foo + bar
  // oops, just wanted one token
  ctx.reset();
  let {value} = ctx.next(); // foo
  //...
}

m foo + bar
```

Other possible names:
- `.clear`
- `.revert`
- `.rewind`

**Comment by disnet** [comment]
Also worth considering along with this (via @vendethiel) is a non-advancing method like `.peek`.

**Comment by gabejohnson** [comment]
I read through #529. Would it make sense to provide a generator as the argument to the transform function making your example:
```
syntax m = ctxGen => {
  let ctx = ctxGen();
  ctx.next('expr'); // foo + bar
  // oops, just wanted one token
  ctx = ctxGen();
  let {value} = ctx.next(); // foo
  //...
}

m foo + bar
```

I wouldn't think this too difficult as you're using persistent data structures for `stxl` in `Enforester`.

**Comment by disnet** [comment]
Yeah, I thought about that too. Definitely worth considering. Maybe something like:

```js
syntax m = ctx => {
  let name = ctx.name();
  let iter = ctx.iter();
  iter.next('expr');
  let { value } = ctx.iter().next(); // start over

  // once we add infix macros again
  let prev = ctx.prevIter();
  prev.next();
  // ...
}
```

**Comment by gabejohnson** [comment]
Arbitrary lookbehind would be pretty sweet (excuse the pun).

**Comment by gabejohnson** [comment]
Have you thought about using a zipper instead of an iterator for tree traversal?

```javascript
syntax m = ctx => {
  ctx.val(); // m
  ctx.next().val(); // 2
  ctx.prev().val(); // 1
  let outer = ctx.outer();
  outer.prev().prev().val(); // a
  outer.prev().inner().val(); // undefined?
  ctx.val(); // m
}
function a() { return 1 m 2 };
```

I realize that this would be departure and a large implementation change. Just thinking out loud.

**Comment by gabejohnson** [comment]
@disnet, `outer` might be overkill. The use-case I had in mind was Racket-like submodules. An example taken from https://docs.racket-lang.org/guide/Module_Syntax.html#%28part._submodules%29:

```javascript
module land_animals {
  export const lion = "Leo";
  export const tiger = "Tony";
  export const bear = "Yogi";
  export const mouse = "Mickey";
}
module zoo {
  module sea_animals {
    export const dolphin = "Flipper";
    export const shark = "Jaws"; 
  }
  export { lion, tiger, bear } from "land_animals";
  export * from "sea_animals";
}
import { tiger, dolphin } from "zoo";
tiger // Tony
dolphin // Flipper
```

**Comment by disnet** [comment]
Zippers are cool but in this case we actually want to intentionally limit the kinds of traversal a macro is allowed to do. In particular `outer` is a total no-go. Messes up all kinds of invariants we need to keep for hygiene and macro interactions to make sense. For example, how should an outer/inner macro interact if the inner macro can “see” (and modify) the outer macro? 

I think your submodules example can be handled with local-expand rather than needing an `outer`. 

Also, arbitrary look behind has its own problems we ran into in 0.7. My initial thought was to do one-term look behind which I think is good enough for most use-cases and then we can think carefully through the implications of relaxing that restriction (my gut feeling is we can’t).

**Comment by gabejohnson** [comment]
I had a feeling `outer` was too much. Now I know why. Well I wouldn't mind working on either a `reset` or `iter` implementation.

**Comment by gabejohnson** [comment]
Working through the `iter` implementation, I realized there's the possibility of passing multiple iterators to a syntax template:
```js
syntax m = ctx => #`${ctx.iter()}${ctx.iter()}`;
m 1; // 11
```

**Comment by gabejohnson** [comment]
Another issue I've found is deciding which iterator is used to restore an enforester's properties after syntax transformation:
```js
syntax m = ctx => {
  let brackets = ctx.iter().next().value;
  let bCtx = brackets.inner();
  let iter1 = bCtx.iter();
  let iter2 = bCtx.iter();
  iter1.next();
  iter2.next();
  iter2.next();
  return #`42`;
}
m [1,2,3];
```

What is the enforester created in `inner` restored to after the transform?

**Comment by disnet** [comment]
Resolved with #556.
